src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelGetSemanticInfoTests.cs

```diff
@@ -12325,9 +12325,8 @@ static void f()
             Assert.Equal(ConversionKind.Boxing, semanticInfo.ImplicitConversion.Kind);
 
             Assert.Null(semanticInfo.Symbol);
-            Assert.Equal(CandidateReason.NotCreatable, semanticInfo.CandidateReason);
-            Assert.Equal(1, semanticInfo.CandidateSymbols.Length);
-            Assert.Equal("T", semanticInfo.CandidateSymbols.First().Name);
+            Assert.Equal(CandidateReason.None, semanticInfo.CandidateReason);
```

Comment:
>CandidateReason.None [](http://example.com/codeflow?start=25&length=20)

Why did `CandidateReason` change?

Reply:
Ok, let's think what the expected behavior here is. Because 2 explanations can be taken:
1. `CandidateReason` tells why instead of single `Symbol` we get only candidates as part of `SymbolInfo`. Therefore if there are no symbols to begin with, there can be no reason for candidates, so it should be `CandidateReason.None`
2. We have an error case where we cannot create type with `new`, so `CandidateReason` should be `NotCreatable`

I lean towards the second option because it is consistent with behavior we observe if we get symbol info from `TypeName` inside `new TypeName()`: we get candidate symbol with reason `NotCreatable` instead of a single symbol if the whole object creation expression has errors even in cases when the identifier itself is bound with no ambiguities. But I want your conclusion on this question first

Reply:
In the previous implementation, `GetSymbolInfo()` for `new T()` returned `CandidateReason.NotCreatable` and the type symbol `T`. Returning `NotCreatable` seemed reasonable. Returning a type rather than no symbols since there was no constructor seemed somewhat surprising, but it looks like `SymbolInfoFactory.Create()` typically returns `CandidateReason.None` if there are no symbols. Returning `CandidateReason.NotCreatable` for this scenario seems more appropriate, but will callers expect symbols when the reason is something other than `CandidateReason.None`?

Is there a strong reason to make a change to `CSharpSemanticModel.GetSemanticSymbols()` for these cases?

Reply:
> Is there a strong reason to make a change to `CSharpSemanticModel.GetSemanticSymbols()` for these cases?

I originally marked current behavior as "inconsistent" in https://github.com/dotnet/roslyn/pull/74218. Fred agreed with me. You yourself consider that "surprising". So 3 people, 2 of which are from the roslyn compiler team, see that behavior as abnormal. Isn't that enough to tweak behavior a little bit to much our expectations?

> Returning `CandidateReason.NotCreatable` for this scenario seems more appropriate

> will callers expect symbols when the reason is something other than `CandidateReason.None`

There are 2 expectations here:
1. If a symbol cannot be constructed, the candidate reason should be `NotCreatable`
2. `CandidateReason` can be something other than `None` only if instead of returning a specific symbol we fallback to candidates due to semantic errors

Unfortunatelly, in our case these 2 invariants contradict each other. Picking either breaks one of that expectations. Leaving it as is breaks expectations for returned symbols. Given that in the vast majority of use cases candidate reason is not even checked (users mostly need symbols themselves), I think changing behavior in the way, which makes returned symbols more expected in an expense of making candidate reason a bit surprising in one way or another is justified. I also think we should not interfere into `SymbolInfoFactory.Create()` logic and leave `None` as a final candidate reason.

Reply:
I'm not familiar enough with use of `SemanticModel.GetSymbolInfo()` to know if one approach should be preferred over the other, although from the documentation on `CandidateReason`, it seems that `NotCreatable` was at least initially intended for unconstrained type parameters. I will defer to other reviewers on this.